src/compute-client/src/controller/instance.rs

```diff
@@ -2373,9 +2115,33 @@ struct ReplicaCollectionState<T> {
     as_of: Antichain<T>,
     /// Tracks hydration state for this collection.
     hydration_state: HydrationState,
+    /// Read holds on storage inputs to this collection.
+    ///
+    /// These read holds are kept to ensure that the replica is able to read from storage inputs at
+    /// all times it hasn't read yet. We only need to install read holds for storage inputs since
+    /// compaction of compute inputs is implicitly held back by Timely/DD.
+    input_read_holds: Vec<ReadHold<T>>,
```

Comment:
I also like the shorter name, just because it's shorter! Hopefully most readers don't have to concern themselves about what read holds are in here exactly.

> I'm not sure I understand what exactly makes it not necessary here if we do it there.

Read holds on compute collections (i.e. indexes) generally need to be taken when you want to read that index at some time. Except that Timely/DD ensures that active readers of an arrangement can continue to read, so you don't need to take extra read holds just for ensuring that. This is the only thing the replica input read holds want to ensure, so we can skip them for compute collections. In fact, it would be bad to install them because it'd mean that a single slow replica could hold back the read frontier of an index, causing it to grow in memory size on all other active replicas. That happens for storage collections too, but is much less problematic because their history isn't kept in memory.

All the read holds we do explicitly take in the compute controller on compute inputs fulfill some other purpose than allowing installed dataflows to continue reading from indexes. I think currently `compute_dependencies` are the only such case. We need them to keep the `since`s of depended-on indexes far enough back that we can spin up a new replica and hydrate our dataflow from its current `since` frontier. When we spin up a new replica we also re-install the depended-on indexes on that replica, so we don't need the read holds to ensure we can physically read the arrangements on existing replicas. But we need it to (a) hold back the dataflow `as_of`s in the `ComputeCommandHistory` and (b) hold back the transitive storage dependencies of those compute inputs.

Reply:
Ok, thanks for the explanation!